### PR TITLE
Add SOTP feature, implemented in C++

### PR DESCRIPTION
**Pull request by davidsaada**

### CODE REVIEW

File: features/FEATURE_SOTP/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# SOTP
```
Comment (dannybenor) [inline features/FEATURE_SOTP/README.md]: Need to add some documentation about how to use it (singletone, etc)

File: features/FEATURE_SOTP/mbed_lib.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  "config": {
+    "thread_safe": {
+        "macro_name": "SOTP_THREAD_SAFE",
+        "value": 0,
+        "help": "Is SOTP code thread safe"
+    },
+    "num_types": {
+        "macro_name": "SOTP_MAX_TYPES",
+        "value": 16,
+        "help": "Number of SOTP types"
```
Comment (dannybenor) [inline features/FEATURE_SOTP/mbed_lib.json]: Change the help to "Number of different keys allowed"

File: features/FEATURE_SOTP/mbed_lib.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    },
+    "area_1_address": {
+        "macro_name": "SOTP_INT_FLASH_AREA_1_ADDRESS",
+        "value": "0xfe000",
+        "help": "Address of SOTP area 1"
+    },
+    "area_1_size": {
+        "macro_name": "SOTP_INT_FLASH_AREA_1_SIZE",
+        "value": "0x1000",
+        "help": "Size of SOTP area 1"
```
Comment (dannybenor) [inline features/FEATURE_SOTP/mbed_lib.json]: Size in bytes of...

File: features/FEATURE_SOTP/source/sotp.cpp [hunk 4]
```diff
@@ -0,0 +1,1033 @@
...
+
+#include "sotp_os_wrapper.h"
+#include "sotp_int_flash_wrapper.h"
+#include <string.h>
+#include <stdio.h>
+#include <stddef.h>
+
+// --------------------------------------------------------- Definitions ----------------------------------------------------------
+
+#define TRACE_GROUP                     "sotp"
```
Comment (dannybenor) [inline features/FEATURE_SOTP/source/sotp.cpp]: Not needed anymore if we do not use mbedtrace

File: features/FEATURE_SOTP/source/sotp.cpp [hunk 5]
```diff
@@ -0,0 +1,1033 @@
...
+       crc = crc ^ (uint32_t) (data_buf[i]);
+       for (j = 0; j < 8; j++) {
+          mask = -(crc & 1);
+          crc = (crc >> 1) ^ (0xEDB88320 & mask);
+       }
+    }
+    return crc;
+}
+
+#if 0
```
Comment (dannybenor) [inline features/FEATURE_SOTP/source/sotp.cpp]: Please remove

File: features/FEATURE_SOTP/source/sotp.h [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+ * @brief Set number of types.
+ *
+ * @returns None.
+ */
+    void set_num_types(uint8_t in_num_types);
+
+/**
+ * @brief Returns one item of data programmed on Flash, given type.
+ *
+ * @param [in] type
```
Comment (paul-szczepanek-arm) [inline features/FEATURE_SOTP/source/sotp.h]: -please explain what type means, I have no clue, maybe this should be an enum?-

I just figured it out, type is the handle/key to data
maybe type is not a very self-explanatory term for this

**Pull request closed without merging**